### PR TITLE
🐛 Prevent local presence transformation when presence created late

### PR DESCRIPTION
**Pull request by alecgibson**
This change fixes a presence edge case:

 1. Register a `doc.on('op')` listener which submits presence
 2. Create the `LocalPresence` _after_ registering this listener
 3. Submit an op
 4. The presence being submitted is transformed against the submitted op
    when it shouldn't be

This happens because the `LocalPresence` handles its transformation
logic in the `'op'` event. If the `LocalPresence` is created after
another handler that submits presence, then its listener is bound late,
and is only triggered after the handler that submits the op.

The result is that the `LocalPresence` will try to transform the
presence, even though the presence is already consistent with the doc's
state.

We could in theory try to prepend the `LocalPresence` listener, but we'd
still be vulnerable to the same bug if consumers _also_ prepend
listeners, and we just move the problem to another place.

Instead, this change fixes the issue by tracking the state of `doc.data`
when the presence is submitted.

However, since `doc.version` is only incremented when an op is
[acknowledged][1], we can't use this to track our document state. To
address this, this change adds a new `_dataStateVersion` internal
variable, which is incremented every time `doc.data` is changed. This
allows us to compare the state of our local `doc.data` over time and
we can use this to check if our presence needs transformation or not.

[1]: https://github.com/share/sharedb/blob/a1a33852531432baeb3cb948ca36500bdbc60fa4/lib/client/doc.js#L947-L948

### CODE REVIEW

**Comment by curran** [comment]
Wow nice catch! Very subtle.

File: lib/client/doc.js [hunk 1]
```diff
@@ -648,11 +659,13 @@ Doc.prototype._otApply = function(op, source) {
...
-        this.type.createDeserialized(op.create.data) :
-        this.type.deserialize(this.type.create(op.create.data)) :
-      this.type.create(op.create.data);
+    this._setData(
+      (this.type.deserialize) ?
+        (this.type.createDeserialized) ?
+          this.type.createDeserialized(op.create.data) :
+          this.type.deserialize(this.type.create(op.create.data)) :
+        this.type.create(op.create.data)
+    );
```
Comment (dawidreedsy) [inline lib/client/doc.js:668]: Can we change this to ifs as it looks more readable 

```suggestion
let data;

if (this.type.deserialize) {
     if (this.type.createDeserialized) {
       data = this.type.createDeserialized(op.create.data)
     } else {
       data = this.type.deserialize(this.type.create(op.create.data))
     }
} else {
     data = this.type.create(op.create.data)
}
this._setData(data);
```

File: lib/client/presence/local-doc-presence.js [hunk 2]
```diff
@@ -76,9 +81,14 @@ LocalDocPresence.prototype._registerWithDoc = function() {
 
 LocalDocPresence.prototype._transformAgainstOp = function(op, source) {
   var presence = this;
+  var docDataVersion = this._doc._dataStateVersion;
+
   this._pendingMessages.forEach(function(message) {
+    var messageDocDataVersion = presence._docDataVersionByPresenceVersion[message.pv];
```
Comment (ericyhwang) [inline lib/client/presence/local-doc-presence.js:87]: Would be nice to add a brief comment about what case this is trying to handle.

**Review by ericyhwang — COMMENTED** [review]
We also discussed adding an internal "_preOp" event, and having presence hook into that. It would be easier to reason about and passes the new test cases, but we're worried about:
- Future internal code may use the "_preOp" event and break the presence transform edge cases
- If we want to extract presence into a plugin in the future (once sharedb supports plugins), then depending on an internal event is a no-go.

**Review by ericyhwang — APPROVED** [review]
